_posts/2016-03-01-Docker-meets-Continuous-Deployment.md

```diff
@@ -0,0 +1,179 @@
+
+## Set The Scene
+
+From developers perspective any application is manifested by a code. There is long way before we finally find some place on a production environment.
+To show this process we need small application. Our application is going to expose some HTTP API which we are going to use after the deployment. 
+Let's assume that will are going to use **Gradle** to build the application and **TeamCity** as build automation tool.
+
+On each **TeamCity** build agent we should have installed **Docker**. We will also use this machine to deploy our application.
+In the real project probably you don't want to do that, but the deployment phase can be changed to fit better to your environment.
+ 
+## Build Application
+
+We need to build something before we will be able to create any docker image. To do that we need to create the first step in our Continuous Deployment (CD) pipeline. 
+In this step we will download our source code, perform all **test** which can be done on the source code level and produce **artifact** which 
+contains all elements required to start and run our application.
+
+This build configuration is not much different then usual step in a common Continuous Deployment pipeline.
+Alongside common parameters we have to define **artifacts** which will be generated after each execution. We are going to use them as base for next steps in the pipeline 
+
+In **TeamCity** we define **artifacts** by defining paths to files from working directory (which is created for each run of a configuration).
+Working directory is combination of files downloaded from a version control system and files generated during execution of build steps.
+We will define later elements of this directory. 
+
+We need two artifacts. The first one is `Dockerfile`. We already prepared this file in our source code and it will be stored in `docker` directory.
+The next file is ***tar*** file which will be generated by our **Gradle** build. It contains script which allows to execute our code and all required libraries. 
+
+<img src="/assets/img/custom/blog/2016-03-01-docker-meets-continuous-deployment/build_config.png" alt="Build Configuration" title="Build Configuration" class="img img-responsive style-screengrab">
+
+Now we are ready to instruct our build configuration how to download our source code. We use github as our code repository.
+We just have to choose type as *Git* and provide url to our application.
+
+<img src="/assets/img/custom/blog/2016-03-01-docker-meets-continuous-deployment/build_vcs.png" alt="Version Control Configuration" title="Version Control Configuration" class="img img-responsive style-screengrab">
+
+We have the source code, so we can describe what to do with this code. As I mention before we will **test** our application and generate an **artifact** with executable content.   
+
+**TeamCity** have predefined runners for different build tools. In this case we are going to use **Gradle** runner to **test** and **build**. 
```

Comment:
TeamCity has predefined runners

In this case we are going to use the Gradle runner